src/Worker.Extensions.DurableTask/DurableTaskExtensionStartup.cs

```diff
@@ -26,13 +32,34 @@ public override void Configure(IFunctionsWorkerApplicationBuilder applicationBui
         applicationBuilder.Services.AddOptions<DurableTaskClientOptions>()
             .PostConfigure<IServiceProvider>((opt, sp) =>
             {
-                if (sp.GetService<DataConverter>() is DataConverter converter
-                    && !ReferenceEquals(opt.DataConverter, JsonDataConverter.Default))
+                if (GetConverter(sp) is DataConverter converter)
                 {
                     opt.DataConverter = converter;
                 }
             });
 
+        applicationBuilder.Services.AddOptions<DurableTaskWorkerOptions>()
+            .PostConfigure<IServiceProvider>((opt, sp) =>
+            {
+                if (GetConverter(sp) is DataConverter converter)
+                {
+                    opt.DataConverter = converter;
+                }
+            });
+
+        applicationBuilder.Services.TryAddSingleton(sp =>
+        {
+            DurableTaskWorkerOptions options = sp.GetRequiredService<IOptions<DurableTaskWorkerOptions>>().Value;
+            ILoggerFactory factory = sp.GetRequiredService<ILoggerFactory>();
+            return new DurableTaskShimFactory(options, factory); // For GrpcOrchestrationRunner
+        });
+
         applicationBuilder.UseMiddleware<DurableTaskFunctionsMiddleware>();
     }
+
+    private static DataConverter? GetConverter(IServiceProvider services)
+    {
+        WorkerOptions? worker = services.GetRequiredService<IOptions<WorkerOptions>>()?.Value;
+        return worker?.Serializer is ObjectSerializer serializer ? new ObjectConverterShim(serializer) : null;
```

Comment:
Is the expression `worker?.Serializer is ObjectSerializer serializer` simply a null check, or is there a possibility that some other type besides `ObjectSerializer` could be configured? This is one of the things I don't like about modern C# language features - it's unclear to me what the actual intent is.

Reply:
it's a null check here. Wound up like that after some refactors, but I see how a regular `is null` would be better. I'll change it.